### PR TITLE
mwan3: Packet Loss & Latency - ubus integration

### PR DESCRIPTION
**Pull request by codemarauder**
Maintainer: @feckert 
Compile tested: No
Run tested: x86, geode, 17.01.4

Description:
1. Link check with pre-defined thresholds for packate loss and latency
2. Extended ubus integration to provide packet loss and latency values per wan per track_ip

Signed-off-by: Nishant Sharma <codemarauder@gmail.com>

### CODE REVIEW

**Comment by codemarauder** [comment]
@feckert rebased the changes against your latest commit. Please test.

**Comment by feckert** [comment]
@codemarauder thanks for your contribution will test and comment it next week. ( I am on vacation thus week )

**Comment by codemarauder** [comment]
@feckert Hope you had a nice vacation. 

Any luck with the testing? I am running it in production for more than a while now.

**Comment by feckert** [comment]
@codemarauder LGTM

One question for clarification for lines 
https://github.com/openwrt/packages/pull/5951/files#diff-87268a842bab5effbe181f9f18b2d348R152
https://github.com/openwrt/packages/pull/5951/files#diff-87268a842bab5effbe181f9f18b2d348R161
https://github.com/openwrt/packages/pull/5951/files#diff-87268a842bab5effbe181f9f18b2d348R171

Why do we have this window check?
Is this for the possibility if the quality of the connection is flapping?

**Comment by codemarauder** [comment]
@feckert 

The first two lines are to ensure that link is considered up only when *both* the latency and packet loss is between the upper and lower watermarks. Link is considered down when *either* of the latency or packet loss breaches the upper watermark.

The flapping condition is taken care by up/down and interval values from the mwan3 config. 
Keeping resultant values of up*interval and down*interval high enough (e.g. 60 seconds) helps.

3rd line is kept as it is from your code where it was writing "skipped" to the status files.

File: net/mwan3/files/etc/config/mwan3 [hunk 1]
```diff
@@ -13,6 +14,10 @@ config interface 'wan'
 	option reliability '2'
 	option count '1'
 	option timeout '2'
+	option max_latency '308'
```
Comment (feckert) [inline net/mwan3/files/etc/config/mwan3]: Could we set this to 300?

File: net/mwan3/files/usr/sbin/mwan3track [hunk 2]
```diff
@@ -86,13 +86,21 @@ main() {
 	config_get failure_interval $1 failure_interval $interval
 	config_get_bool keep_failure_interval $1 keep_failure_interval 0
 	config_get recovery_interval $1 recovery_interval $interval
+	config_get_bool check_quality $1 check_quality 0
+	config_get max_latency $1 max_latency $max_latency
+	config_get min_latency $1 min_latency $min_latency
+	config_get max_loss $1 max_loss $max_loss
+	config_get min_loss $1 min_loss $min_loss
```
Comment (feckert) [inline net/mwan3/files/usr/sbin/mwan3track]: Could we also load the defaults if no max/min values are found in the config?

**Comment by feckert** [comment]
> The flapping condition is taken care by up / down and interval values from the mwan3 config.
Keeping resultant values of up * interval and down * interval high enough (e.g. 60 seconds) helps.

If so why do we need a **min** values is it not enough only use a max value?
if either loss or latency is over the max values then link quality is bad and so lost++ is executed.
In all other state the value is good because it is below the max value?

**Comment by codemarauder** [comment]
On Monday 07 May 2018 03:43 PM, Florian Eckert wrote:
> 
> If so why do we need a *min* values is it not enough only use a max value?

As Latency and Loss are not boolean like ping, we actually need a range. 
Let me try to explain below.

> if either loss or latency is over the max values then link quality is 
> bad and so lost++ is executed.
> In all other state the value is good because it is below the max value?

Well, then there will be a problem if loss reduces from 30%+ (max_loss) 
to 29% or latency reduces from 300ms+ (max_latency) to 290ms, but still 
link is not usable, as we expect loss to be <5% and base latency of the 
link is usually <50ms.

We set lower thresholds for acceptable link quality level.

**Comment by feckert** [comment]
@codemarauder thanks for clarification but if so could we change the the configuration values from

max_latency -> failure_latency
max_loss ->failure_loss
min_latency -> recovery_latency
min_loss -> recovery_loss

This would be clarify the values and would be consistent with **failure_interval** and **recovery_interval**

If this is change we could merge this

Thanks

**Comment by feckert** [comment]
@codemarauder Please also update the PKG_VERSION to 2.6.15

File: net/mwan3/files/usr/sbin/mwan3track [hunk 3]
```diff
@@ -109,23 +117,58 @@ main() {
 			if [ $host_up_count -lt $reliability ]; then
 				case "$track_method" in
 					ping)
-						ping -I $DEVICE -c $count -W $timeout -s $size -q $track_ip &> /dev/null ;;
+							if [ $check_quality -eq 0 ]; then
+								ping -I $DEVICE -c $count -W $timeout -s $size -q $track_ip &> /dev/null
```
Comment (feckert) [inline net/mwan3/files/usr/sbin/mwan3track]: @codemarauder during testing i have noticed a problem!

The check in line 138 is always success and system will not detect a fault!
On my system if apply your changes and use only ping without quality check the ping check is always successful even though the target ping could not get reached!

Do fix this we have to save the **$?** into a variable for example **$result** on **ping/arping/httping** and then change the check from **if [ $? -eq 0 ]; then** to **if [ $result -eq 0 ]; then**
this should fix the problem.

**Comment by feckert** [comment]
@codemarauder I want to get this feature into the next openwrt stable which is on the road because it is a nice feature :-)
I would suggest to cherry-pick your changes and add some additional fixes on top your changes.
I will not touch your commit so your name will be mention?
Is this OK for you?

File: net/mwan3/files/usr/sbin/mwan3track (hunk 2 above)
Comment (codemarauder) [reply net/mwan3/files/usr/sbin/mwan3track]: Let's try to identify some sane / safe defaults.
For packetloss, <5% and >20% should work.
For latency, it is difficult as monitor IP (openDNS or google DNS) latency would differ across the globe, unless user checks the base latency and configures mwan3 properly. 

<500ms and >1000ms should be safe, although not sane.

Your views?

File: net/mwan3/files/usr/sbin/mwan3track (hunk 3 above)
Comment (codemarauder) [reply net/mwan3/files/usr/sbin/mwan3track]: Got it!
I recall why I had to use $ping_result, as $? was always true for me as well. I somehow missed that other methods are also using $? and need to be modified.

I will try to fix things tonight.

**Comment by codemarauder** [comment]
@feckert As suggested by you, I have made the changes and updated the pull request with further commits. 

Thanks a lot for testing the modifications and encouraging me to go ahead with further commits.

Looking forward for further test results.

**Comment by feckert** [comment]
@codemarauder  could you please close this PR I will open another one with your initial commit and add some additional fixes on top of this?

**Comment by codemarauder** [comment]
@feckert Closing it. I am not very well versed with git, so :)

Cheers!

**Comment by feckert** [comment]
@codemarauder no problem
Thanks for your contribution
I have opened a new pull request with your feature and fixes on top by me.
See pull request https://github.com/openwrt/packages/pull/6036

**Pull request closed without merging**